src/allocation_engine/interval_tree.rs

```diff
@@ -359,6 +361,106 @@ impl InnerNode {
         self.update_cached_height();
         self.rotate()
     }
+
+    /// Insert a new node in the subtree. After the node is inserted the
+    /// tree will be balanced.
+    #[allow(dead_code)]
+    pub(crate) fn insert(
+        mut self: Box<Self>,
+        key: Range,
+        node_state: NodeState,
+    ) -> Result<Box<Self>> {
+        if self.key.overlaps(&key) && self.key != key {
+            return Err(Error::Overlap(key, self.key));
+        }
+        match self.key.cmp(&key) {
+            Ordering::Equal => {
+                return Err(Error::ResourceExhausted);
+            }
+            Ordering::Less => match self.right {
+                None => self.right = Some(Box::new(InnerNode::new(key, node_state))),
+                Some(_) => {
+                    self.right = self
+                        .right
+                        .take()
+                        .map(|n| n.insert(key, node_state).unwrap())
```

Comment:
We should return an error here instead of unwrapping. There may be a case where a duplicate node is being inserted at a lower level in the subtree, which will cause a panic in this case instead of gracefully returning.